File: CMPS3500_Minischeme_Group08/procedural/recursion.cpp
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga  /
  / ASGT: CHECKPOINT 3                                   /
  / ORGN: CSUB - CMPS 3500                               /
  / FILE: recursion.cpp                                  /
  / DATE: 05/1/2026                                      / 
*/
#include "recursion.h"
#include "evaluate.h"
#include "scope.h"

bool isRecursiveDefinition(const std::vector<std::string>& value_expr)
{
    return !value_expr.empty() && value_expr[0] == "(" &&
           value_expr.size() > 1 && value_expr[1] == "lambda";
}

std::string handleRecursiveDefine(
    const std::string& name,
    const std::vector<std::string>& value_expr,
    Scope* scope
)
{
    std::string value = evaluate(value_expr, scope);
    addScopeEntry(scope, name, value);
    return "";
}

File: CMPS3500_MiniScheme_Starter_Package/oop/Recursion.java
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga /
  / ASGT: CHECKPOINT 3                                  /
  / ORGN: CSUB - CMPS 3500                              /
  / FILE: Recursion.java                                /
  / DATE: 05/01/2026                                    /
*/

import java.util.List;

public class Recursion {
    public static boolean isRecursiveDefinition(List<String> valueExpr){
        return !valueExpr.isEmpty() &&
        valueExpr.get(0).equals("(") &&
        valueExpr.size() > 1 &&
        valueExpr.get(1) == "lambda";
    }

    public static String handleRecursiveDefine(
        String name,
        List<String> valueExpr,
        Scope scope
    ){
        String value = Evaluate.evaluate(valueExpr, scope);
        scope.addScopeEntry(name, value);
        return "";
    }
}
